tests/federation_room_join_partial_state_test.go

```diff
@@ -1229,6 +1229,125 @@ func TestPartialStateJoin(t *testing.T) {
+
+		if err == nil {
+			t.Errorf("MakeKnock returned 200, want 404")
+		} else if httpError, ok := err.(gomatrix.HTTPError); ok {
+			t.Logf("MakeKnock => %d/%s", httpError.Code, string(httpError.Contents))
+			if httpError.Code != 404 {
+				t.Errorf("expected 404, got %d", httpError.Code)
+			}
+			errcode := must.GetJSONFieldStr(t, httpError.Contents, "errcode")
+			if errcode != "M_NOT_FOUND" {
+				t.Errorf("errcode: got %s, want M_NOT_FOUND", errcode)
+			}
+		} else {
+			t.Errorf("MakeKnock: non-HTTPError: %v", err)
+		}
+	})
+
+	// when the server is in the middle of a partial state join, it should not accept
+	// /send_knock because it can't give a full answer.
+	t.Run("Rejects send_knock during partial join", func(t *testing.T) {
+		// In this test, we have 3 homeservers:
+		//   hs1 (the server under test) with @alice:hs1
+		//     This is the server that will be in the middle of a partial join.
+		//   testServer1 (a Complement test server) with @charlie:<server name>
+		//     This is the server that will create the room originally.
+		//   testServer2 (another Complement test server) with @daniel:<server name>
+		//     This is the server that will try to knock on the room via hs2,
+		//     but only after using hs1 to /make_knock (as otherwise we have no way
+		//     of being able to build a request to /send_knock)
+		//
+		deployment := Deploy(t, b.BlueprintAlice)
+		defer deployment.Destroy(t)
+		alice := deployment.Client(t, "hs1", "@alice:hs1")
+
+		testServer1 := createTestServer(t, deployment)
+		cancel := testServer1.Listen()
+		defer cancel()
+		serverRoom := createTestRoom(t, testServer1, alice.GetDefaultRoomVersion(t))
+		psjResult := beginPartialStateJoin(t, testServer1, serverRoom, alice)
+		defer psjResult.Destroy()
+
+		// hs1's partial join is now in progress.
+		// Let's have a test server rock up and ask to /send_knock in the room via hs1.
+		// To do that, we need to /make_knock first.
+		// Asking hs1 to /make_knock won't work, because it should reject that request.
+		// To work around that, we /make_knock via hs2.
+
+		testServer2 := createTestServer(t, deployment)
+		cancel2 := testServer2.Listen()
+		defer cancel2()
+
+		fedClient2 := testServer2.FederationClient(deployment)
+
+		// Manually /make_knock via testServer1.
+		// This is permissible because testServer1 is fully joined to the room.
+		// We can't actually use /make_knock because host.docker.internal doesn't resolve,
+		// so compute it without making any requests:
+		makeKnockResp, err := federation.MakeRespMakeKnock(testServer1, serverRoom, testServer2.UserID("daniel"))
+		if err != nil {
+			t.Fatalf("MakeRespMakeKnock failed : %s", err)
+		}
+
+		// daniel then tries to /send_knock via the homeserver under test
+		knockEvent, err := makeKnockResp.KnockEvent.Build(time.Now(), gomatrixserverlib.ServerName(testServer2.ServerName()), testServer2.KeyID, testServer2.Priv, makeKnockResp.RoomVersion)
+		must.NotError(t, "KnockEvent.Build", err)
```

Comment:
is this just shorthand for `if err != nil { /* fail the test with given message */ }` ?

Reply:
yeah, seems to be
```go
// NotError will ensure `err` is nil else terminate the test with `msg`.
func NotError(t *testing.T, msg string, err error) {
	t.Helper()
	if err != nil {
		t.Fatalf("MustNotError: %s -> %s", msg, err)
	}
}
```